pytest_inmanta/plugin.py

```diff
@@ -489,26 +464,51 @@ def _create_project_and_load(self, model: str) -> module.Project:
         * Install the module dependencies
         * Load the project
 
+        :param init: True iff the project should start from a clean slate. Ignored for older (<6) versions of core.
         :return: The newly created module.Project instance.
         """
+        # update collection of all registered plugins before executing side effects
+        self._registered_plugins.update(plugins.PluginMeta.get_functions())
+
         with open(os.path.join(self._test_project_dir, "main.cf"), "w+") as fd:
             fd.write(model)
 
-        signature: inspect.Signature = inspect.Signature.from_callable(
+        signature_init: inspect.Signature = inspect.Signature.from_callable(
             module.Project.__init__
         )
         # The venv_path parameter only exists on ISO5+
-        extra_kwargs = (
+        extra_kwargs_init = (
             {"venv_path": self._env_path}
-            if "venv_path" in signature.parameters.keys()
+            if "venv_path" in signature_init.parameters.keys()
             else {}
         )
-        test_project = module.Project(self._test_project_dir, **extra_kwargs)
-        module.Project.set(test_project)
+        test_project = module.Project(self._test_project_dir, **extra_kwargs_init)
+        signature_set: inspect.Signature = inspect.Signature.from_callable(
+            module.Project.set
+        )
+        # For supported versions of core, don't clean up loaded modules between invocations to keep top-level imports valid
+        extra_kwargs_set = (
+            {"clean": False} if "clean" in signature_set.parameters.keys() else {}
+        )
+        module.Project.set(test_project, **extra_kwargs_set)
+
+        # The compiler expects plugins to be registered only for modules that are currently loaded
+        #   => deregister plugins, load the project, then complete the set from the previously registered ones
+        #   Completing the set is required because we didn't allow the project to do a cleanup, meaning it is not guaranteed to
+        #   register previously registered plugins (and indeed in practice it does not).
+        plugins.PluginMeta.clear()
         if hasattr(test_project, "install_modules"):
             # more recent versions of core require explicit modules installation (ISO5+)
             test_project.install_modules()
         test_project.load()
+        currently_registered_plugins: typing.Mapping[str, typing.Type[plugins.Plugin]] = plugins.PluginMeta.get_functions()
+        loaded_mod_ns_pattern: typing.Pattern[str] = re.compile(
+            "(" + "|".join(re.escape(mod) for mod in test_project.modules.keys()) + ")::"
+        )
+        for fq_plugin_name, plugin in self._registered_plugins.items():
+            if fq_plugin_name not in currently_registered_plugins and loaded_mod_ns_pattern.match(fq_plugin_name):
+                plugins.PluginMeta.add_function(plugin)
```

Comment:
I see, this works because `project.load()` builds the ast, but plugins are registered in in 'compiler.compile`

Reply:
Almost. Plugins are registered in `project.load()` as well, but they are added to the namespace only in `compiler.compile`. So by reregistering them after `project.load()`, we mimic the normal behavior.

Reply:
Almost. Plugins are registered in `project.load()` as well, but they are added to the namespace only in `compiler.compile`. So by reregistering them after `project.load()`, we mimic the normal behavior.